Updated the interface documentation (#753)
PennyLaneAI/pennylane: doc/introduction/interfaces/numpy.rst

The :ref:`QNode decorator <intro_vcirc_decorator>` is the recommended way for creating QNodes
in PennyLane. By default, all QNodes are constructed for the NumPy interface,
but this can also be specified explicitly by passing the ``interface='autograd'`` keyword argument:

.. code-block:: python

    dev = qml.device('default.qubit', wires=2)

>     @qml.qnode(dev, interface='autograd')

Author: (docstrings have long favoured using `'autograd'` over `'numpy'`, so I've made the update here as well)

    phi = np.array([0.5, 0.1])
    theta = 0.2
    dcircuit = qml.grad(circuit3)

Evaluating this gradient function at specific parameter values:

>>> dcircuit(phi, theta)
(array([ -4.79425539e-01,   1.11022302e-16]), array(0.0))

> Non-differentiable arguments
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> Note that **all positional arguments provided to the QNode are assumed to be differentiable

Reviewer: ```suggestion
Differentiable and non-differentiable arguments
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

How does PennyLane know which arguments of a quantum function are differentiable and which ones are not? 

As a basic rule, **all positional arguments provided to the QNode are assumed to be differentiable
```

> by default**:

Reviewer: ```suggestion
by default, while keyword arguments are not.**:
```

Reviewer: Is that even true?

Author: Keyword arguments can never be differentiable in the current system :) I would like to change that, so that the type of argument no longer affects differentiability at all (hence why I never mentioned keyword args here, I did that somewhat deliberately)

> >>> from pennylane import numpy as np
> >>> np.array([0.1, 0.2])
> tensor([0.1, 0.2], requires_grad=True)

Reviewer: ```suggestion

```
However, you may want to pass arguments as positional arguments to a QNode but _not_ have PennyLane
consider them when computing gradients. To accomplish this, it is important to know that importing numpy from PennyLane has an important side effect: creating of arrays with ``np.array()`` produces 
a _tensor_ which behaves like "normal" numpy arrays, but has a special flag indicating whether it is trainable or not. 

>>> from pennylane import numpy as np
>>> np.array([0.1, 0.2])
tensor([0.1, 0.2], requires_grad=True)
```

> If you would like to provide explicit non-differentiable arguments to the
> QNode or gradient function, make sure to use a NumPy array that specifies
> ``requires_grad=False``:

Reviewer: ```suggestion
If you would like to provide explicit non-differentiable arguments to the
QNode or gradient function, make sure to import numpy via ``from pennylane import numpy as np``
and create a NumPy array that specifies
``requires_grad=False``:
```

Author: Have rewritten this section to include your changes (and have added the import line to the examples to make it explicit!)

>>> np.array([0.1, 0.2], requires_grad=False)
tensor([0.1, 0.2], requires_grad=False)

> The ``requires_grad`` argument can be passed to any NumPy function provided by
> PennyLane.

Reviewer: ```suggestion
.. note::

  The ``requires_grad`` argument can be passed to any NumPy function provided by PennyLane.
```

For example, consider the following QNode that accepts one trainable argument ``weights``,
and two non-trainable arguments ``data`` and ``wires``:

.. code-block:: python

    dev = qml.device('default.qubit', wires=5)

    @qml.qnode(dev)
    def circuit(weights, data, wires):
        qml.templates.AmplitudeEmbedding(data, wires=wires, normalize=True)
        qml.RX(weights[0], wires=wires[0])
        qml.RY(weights[1], wires=wires[1])

>         qml.RZ(weights[2], wires=wires[2])

Reviewer: ```suggestion
```

>         qml.CNOT(wires=[wires[0], wires[1]])

Reviewer: ```suggestion
```

Just to shorten?

Author: Ah, the reason I added these extra gates was because I kept getting trivial Jacobians otherwise 😆

        qml.CNOT(wires=[wires[0], wires[2]])
        return qml.expval(qml.PauliZ(wires[0]))

We must specify that ``data`` and ``wires`` are NumPy arrays with ``requires_grad=False``:

>>> weights = np.array([0.1, 0.2, 0.3])
>>> data = np.random.random([2**3], requires_grad=False)
>>> wires = np.array([2, 0, 1], requires_grad=False)
>>> circuit(weights, data, wires)
0.16935626052294817

> 

Reviewer: Shall we add something along the lines of 

If you do not want to specify the `requires_grad` flag, simply create a QNode that accepts the non-differentiable arguments as keyword arguments:

[Example]

This will also work if you create arrays after importing vanilla numpy with ``import numpy as np``.

Reviewer: Maybe fits better at the end...

If we were to naively try computing the gradient of ``circuit5`` using the :func:`~.grad` function,

.. code::

    g1 = qml.grad(circuit5, argnum=0)
    g1(np.pi/2)
>>> g1 = qml.grad(circuit5)
>>> params = np.array([np.pi/2, 0.2])
>>> g1(params)

we would get an error message. This is because the `gradient <https://en.wikipedia.org/wiki/Gradient>`_ is
only defined for scalar functions, i.e., functions which return a single value. In the case where the QNode
returns multiple expectation values, the correct differential operator to use is
the `Jacobian matrix <https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant>`_.
This can be accessed in PennyLane as :func:`~.jacobian`:

>>> j1 = qml.jacobian(circuit5, argnum=0)
>>> j1(np.pi/2)
array([-1. -1.])
>>> j1 = qml.jacobian(circuit5)
>>> j1(params)
array([[ 0.        , -0.98006658],
       [-0.98006658,  0.        ]])

The output of :func:`~.jacobian` is a two-dimensional vector, with the first/second element being
the partial derivative of the first/second expectation value with respect to the input parameter.
The Jacobian function has the same signature as the gradient function, requiring the user to specify

> which argument should be differentiated.

Author: Removed, as this is no longer the case post-variable refactor

If you want to compute the Jacobian matrix for a function with multiple input parameters
and multiple expectation values, the recommended way to do this is to combine the parameters
into a single list/array and index into this inside your quantum circuit function.
Consider the following circuit:

.. code-block:: python

    @qml.qnode(dev)
    def circuit6(params):
        qml.RX(params[0], wires=0)
        qml.RZ(params[1], wires=0)
        qml.RX(params[2], wires=1)
        qml.CNOT(wires=[0, 1])
        return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))

It has a full Jacobian with two rows and three columns:

>>> j2 = qml.jacobian(circuit6, argnum=0)
>>> params = np.array([np.pi / 3, 0.25, np.pi / 2])
>>> j2(params)
array([[-8.66025404e-01, -5.55111512e-17,  0.00000000e+00],
       [-4.71844785e-16, -1.38777878e-17, -5.00000000e-01]])

.. warning::

    Currently, :func:`pennylane.jacobian` supports only the case where
    ``argnum`` is a single integer. For quantum functions with multiple arguments,

>     use the above method of combining arguments in an array to get the full Jacobian matrix.

Reviewer: Nice!

> The PennyLane NumPy interface leverages the Python library `autograd

Author: This section used the term 'user', and it shouldn't, so I gave it a quick re-write
